// angular-azure-blob-service/esm5/src/app/modules/blob/blob.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { BlobService } from './blob.service';
export { BlobService } from './blob.service';
var BlobModule = /** @class */ (function () {
    function BlobModule() {
    }
    /**
     * @return {?}
     */
    BlobModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: BlobModule,
            providers: [
                BlobService
            ]
        };
    };
    BlobModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule
                    ],
                    declarations: [],
                    exports: []
                },] }
    ];
    return BlobModule;
}());
export { BlobModule };
